library/std/src/io/buffered/bufwriter.rs

```diff
@@ -319,6 +333,95 @@ impl<W: Write> BufWriter<W> {
         let buf = if !self.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
         (self.inner.take().unwrap(), buf)
     }
+
+    // Ensure this function does not get inlined into `write`, so that it
+    // remains inlineable and its common path remains as short as possible.
+    // If this function ends up being called frequently relative to `write`,
+    // it's likely a sign that the client is using an improperly sized buffer
+    // or their write patterns are somewhat pathological.
+    #[cold]
+    fn write_cold(&mut self, buf: &[u8]) -> io::Result<usize> {
```

Comment:
Looks like it's assumed this function is only called for `buf.len() >= self.spare_capacity()`. It'd make it easier to read and verify if that assumption was written down. Perhaps as a `debug_assert!()`. (Same below.)

Reply:
It's only called in that case, but it still works if that assumption doesn't hold, if I'm not mistaken, so I didn't add a `debug_assert!`. Let me know if you still think I should add it and I'll go ahead.